### PR TITLE
Enable osx-arm64 builds

### PR DESCRIPTION
**Pull request by jaimergp**
MNT: Re-rendered with conda-build 24.1.2, conda-smithy 3.31.1, and conda-forge-pinning 2024.02.26.11.01.00

Checklist
* [X] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [X] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -27,6 +27,15 @@ requirements:
     - bazel
     - bazel {{ bazel_version }}
     - bazel-toolchain
+    - pip                                 # [target_platform != build_platform]
+    - python                              # [target_platform != build_platform]
+    - cross-python_{{ target_platform }}  # [target_platform != build_platform]
+    - wheel                               # [target_platform != build_platform]
+    - setuptools-scm                      # [target_platform != build_platform]
```
Comment (xhochy) [inline recipe/meta.yaml:34]: Pure-Python packages should not be in the cross-section, only the ones with native code.

Comment (jaimergp) [reply recipe/meta.yaml:34]:
```suggestion
```

**Comment by jaimergp** [comment]
Seems to be working now 🚀

**Comment by traversaro** [comment]
Great, thanks! Just to understand, do you have any idae why do we need all those C++ dependencies as build dependencies? For some reason I can't see the logs.

**Comment by jaimergp** [comment]
I kept adding them one by one as errors popped up. Mostly missing headers, I think? At some point I got tired and added them all 😬 Maybe we can remove a few but it's a time consuming task.

**Comment by traversaro** [comment]
> I kept adding them one by one as errors popped up. Mostly missing headers, I think? At some point I got tired and added them all 😬 Maybe we can remove a few but it's a time consuming task.

Ah, I guess that bazel was using the headers from the build directory instead of host one? That scares me a bit as it could introduce subtle inconsistencies, but to be honest any bazel-using recipe scares me. So I guess it is ok for me, @goanpeca @xhochy feel free to provide additional details.

**Review by traversaro — APPROVED** [review]

**Comment by jaimergp** [comment]
Only for curl. The rest are no longer accessible because Azure kills the log if canceled 🤷 

```
ERROR: /Users/runner/miniforge3/conda-bld/tensorstore_1708969709158/work/python/tensorstore/BUILD:1113:20: Compiling python/tensorstore/metrics.cc failed: (Exit 1): cc_wrapper_build.sh failed: error executing command (from target //python/tensorstore:metrics)
    (cd /Users/runner/miniforge3/conda-bld/tensorstore_1708969709158/_build_env/share/bazel/5865c48d93879b5fd9724b12778b7180/sandbox/darwin-sandbox/863/execroot/tensorstore && \
    exec env - \
      PATH=/Users/runner/miniforge3/conda-bld/tensorstore_1708969709158/_build_env/bin:/Users/runner/miniforge3/conda-bld/tensorstore_1708969709158/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/bin:/Users/runner/miniforge3/condabin:/Users/runner/miniforge3/conda-bld/tensorstore_1708969709158/_build_env/bin:/Users/runner/miniforge3/conda-bld/tensorstore_1708969709158/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/bin:/Users/runner/miniforge3/bin:/Users/runner/miniforge3/condabin:.:/usr/local/bin:/usr/local/sbin:/Users/runner/bin:/Library/Frameworks/Python.framework/Versions/Current/bin:/Library/Frameworks/Mono.framework/Versions/Current/Commands:/usr/bin:/bin:/usr/sbin:/sbin:/Users/runner/.ghcup/bin \
      PWD=/proc/self/cwd \
    bazel_toolchain/cc_wrapper_build.sh -MD -MF bazel-out/darwin-opt/bin/python/tensorstore/_objs/metrics/metrics.pic.d '-frandom-seed=bazel-out/darwin-opt/bin/python/tensorstore/_objs/metrics/metrics.pic.o' -fPIC '-DPYBIND11_NAMESPACE=pybind11' '-DBAZEL_CURRENT_REPOSITORY=""' -iquote . -iquote bazel-out/darwin-opt/bin -iquote external/com_github_pybind_pybind11 -iquote bazel-out/darwin-opt/bin/external/com_github_pybind_pybind11 -iquote external/local_config_python -iquote bazel-out/darwin-opt/bin/external/local_config_python -iquote external/com_google_absl -iquote bazel-out/darwin-opt/bin/external/com_google_absl -iquote external/com_google_boringssl -iquote bazel-out/darwin-opt/bin/external/com_google_boringssl -iquote external/com_google_riegeli -iquote bazel-out/darwin-opt/bin/external/com_google_riegeli -iquote external/com_github_nlohmann_json -iquote bazel-out/darwin-opt/bin/external/com_github_nlohmann_json -iquote external/com_google_re2 -iquote bazel-out/darwin-opt/bin/external/com_google_re2 -iquote external/se_curl -iquote bazel-out/darwin-opt/bin/external/se_curl -Ibazel-out/darwin-opt/bin/external/local_config_python/_virtual_includes/python_headers -isystem /Users/runner/miniforge3/conda-bld/tensorstore_1708969709158/_build_env/include/c++/v1 -isystem /Users/runner/miniforge3/conda-bld/tensorstore_1708969709158/_build_env/lib/clang/16.0.6/include -isystem /Applications/Xcode_13.2.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk/usr/include -isystem /Applications/Xcode_13.2.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk/System/Library/Frameworks -isystem /Users/runner/miniforge3/conda-bld/tensorstore_1708969709158/_build_env/include -ftree-vectorize -fPIC -fstack-protector-strong -O2 -pipe -isystem /Users/runner/miniforge3/conda-bld/tensorstore_1708969709158/_build_env/include '-fdebug-prefix-map=/Users/runner/miniforge3/conda-bld/tensorstore_1708969709158/work=/usr/local/src/conda/tensorstore-0.1.54' '-fdebug-prefix-map=/Users/runner/miniforge3/conda-bld/tensorstore_1708969709158/_build_env=/usr/local/src/conda-prefix' '-D_FORTIFY_SOURCE=2' -isystem /Users/runner/miniforge3/conda-bld/tensorstore_1708969709158/_build_env/include '-mmacosx-version-min=11.0' -ftree-vectorize -fPIC -fstack-protector-strong -O2 -pipe '-stdlib=libc++' -fvisibility-inlines-hidden '-fmessage-length=0' -isystem /Users/runner/miniforge3/conda-bld/tensorstore_1708969709158/_build_env/include '-fdebug-prefix-map=/Users/runner/miniforge3/conda-bld/tensorstore_1708969709158/work=/usr/local/src/conda/tensorstore-0.1.54' '-fdebug-prefix-map=/Users/runner/miniforge3/conda-bld/tensorstore_1708969709158/_build_env=/usr/local/src/conda-prefix' '-D_FORTIFY_SOURCE=2' -isystem /Users/runner/miniforge3/conda-bld/tensorstore_1708969709158/_build_env/include '-mmacosx-version-min=11.0' '-fvisibility=hidden' -c python/tensorstore/metrics.cc -o bazel-out/darwin-opt/bin/python/tensorstore/_objs/metrics/metrics.pic.o)
  # Configuration: 203f2ee1b5179a96a189b54c81691665419dd459df90ce1f82ec9f4a6a8d8ce8
  # Execution platform: @local_config_platform//:host

  Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
  In file included from python/tensorstore/metrics.cc:32:
  In file included from ./tensorstore/internal/http/curl_transport.h:21:
  In file included from ./tensorstore/internal/http/curl_handle.h:19:
  In file included from ./tensorstore/internal/http/curl_factory.h:20:
  ./tensorstore/internal/http/curl_wrappers.h:23:10: fatal error: 'curl/curl.h' file not found
  #include <curl/curl.h>  // IWYU pragma: export
           ^~~~~~~~~~~~~
  1 error generated.
```

**Comment by traversaro** [comment]
The problem seems to be the `-isystem /Users/runner/miniforge3/conda-bld/tensorstore_1708969709158/_build_env/include` flag, that should instead point to the host environment instead of the build environment.

**Comment by xhochy** [comment]
Do you know why you set `CLANG=$CC_FOR_BUILD` in https://github.com/conda-forge/tensorstore-feedstock/blob/f1bfb11cbfac749ab9f9aed79d8968a44eac9ddb/recipe/build.sh#L6 ?

**Comment by xhochy** [comment]
Also https://github.com/conda-forge/tensorstore-feedstock/blob/f1bfb11cbfac749ab9f9aed79d8968a44eac9ddb/recipe/build.sh#L41C25-L41C41 makes no sense, this toolchain doesn't exist and you set the correct options already ~10 lines above. Probably doesn't matter but you should remove that line.

**Comment by xhochy** [comment]
At the moment, I would be sceptical whether we actually get an `osx-arm64` binary and not a `osx-64` one by accident. Either you check that manually or I can look into that in 1-10 days (no good timeline on availability on my side).

**Comment by jaimergp** [comment]
> Also https://github.com/conda-forge/tensorstore-feedstock/blob/f1bfb11cbfac749ab9f9aed79d8968a44eac9ddb/recipe/build.sh#L41C25-L41C41 makes no sense, this toolchain doesn't exist and you set the correct options already ~10 lines above. Probably doesn't matter but you should remove that line.

See this comment in staged-recipes: https://github.com/conda-forge/staged-recipes/pull/17848/files#r1367748543

> Do you know why you set `CLANG=$CC_FOR_BUILD` in
> 
> https://github.com/conda-forge/tensorstore-feedstock/blob/f1bfb11cbfac749ab9f9aed79d8968a44eac9ddb/recipe/build.sh#L6
> 
> ?

Can't remember, so let's try without it :)

**Comment by jaimergp** [comment]
```

  external/blake3/c/blake3_sse41_x86-64_unix.S:33:9: error: unrecognized instruction mnemonic, did you mean: ushl, ushr?
          push r13
          ^
  external/blake3/c/blake3_sse41_x86-64_unix.S:34:9: error: unrecognized instruction mnemonic, did you mean: ushl, ushr?
          push r12
          ^
  external/blake3/c/blake3_sse41_x86-64_unix.S:35:9: error: unrecognized instruction mnemonic, did you mean: ushl, ushr?
          push rbx
          ^
  external/blake3/c/blake3_sse41_x86-64_unix.S:36:9: error: unrecognized instruction mnemonic, did you mean: ushl, ushr?
          push rbp
          ^
  external/blake3/c/blake3_sse41_x86-64_unix.S:37:13: error: invalid operand for instruction
          mov rbp, rsp
              ^
  external/blake3/c/blake3_sse41_x86-64_unix.S:38:13: error: invalid operand for instruction
          sub rsp, 360
              ^
  external/blake3/c/blake3_sse41_x86-64_unix.S:39:13: error: invalid operand for instruction
          and rsp, 0xFFFFFFFFFFFFFFC0
              ^
  external/blake3/c/blake3_sse41_x86-64_unix.S:40:13: error: invalid operand for instruction
          neg r9d
              ^
  external/blake3/c/blake3_sse41_x86-64_unix.S:41:9: error: unrecognized instruction mnemonic, did you mean: fmov, mov, movi, movk, movn, movz, smov, umov?
          movd xmm0, r9d
          ^
...
```

Looks like we are trying to compile some x64 things?

**Comment by jaimergp** [comment]
Woohoo, thank you much @xhochy!

**Comment by xhochy** [comment]
Anything that's stopping this from being merged?

**Review by traversaro — APPROVED** [review]

**Comment by traversaro** [comment]
> Anything that's stopping this from being merged?

Good to go for me!

**Comment by pierreglaser** [comment]
Thank you very much @jaimergp @xhochy @traversaro  for finishing that PR!

**Pull request merged**